# Exploring the Global Summary of the Year Dataset

## First Goal: See how much of the data (.csv files) contains temperature information

Would like to explore trends and answer questions such as: How many countries are seeing increased maximum or average temperatures over time? What is the average trend line of temperature increase? What is the KDE for the trend distribution? Can a dashboard be made to visualize this geographically?
            
            

### Load Libraries

In [101]:
import pandas as pd
import numpy as np
import os
import random
from datetime import datetime
import re

### Define Functions

In [3]:
def percent_missing(dataframe, groupby_col, col_name):
    groupby = dataframe.groupby(dataframe[groupby_col])[col_name].sum()
    proportion_non_na = round(len(groupby[groupby != 0])/len(groupby)*100,2)
    print(f"{proportion_non_na}% of the {groupby_col}s have {col_name} data")

### Convert all 80k+ CSVs into one master CSV (only run once)
   
   #csv_data_files = os.listdir('Global Summary of the Year NCEI Data')
len(csv_data_files)


    #os.chdir('Global Summary of the Year NCEI Data')
df = pd.concat(map(pd.read_csv, csv_data_files), ignore_index=True)

#Change directory 
os.chdir("..")
os.path.abspath(os.curdir)

df.to_csv('GSOY_Master_File.csv')


In [4]:
#Check current working directory

os.path.abspath(os.curdir)

'C:\\Users\\dminkler002\\OneDrive - Guidehouse\\Documents\\Training\\Climate Project\\ClimateProjectGSOY'

### Read in master csv

In [5]:
df_master = pd.read_csv("GSOY_Master_File.csv")

#Preview the CSV
df_master

C:\Users\dminkler002\AppData\Local\Temp\ipykernel_3936\3031860136.py:1: DtypeWarning: Columns (80,82,84,86,88,90,92,94,96,98,100,102,104,106,108,110,112,114,116,118,120,122,124,126,128,130,132,134,136,138,140,142,144,146,148,150,152,154,156,158,160,162,164,166,168,170,172,174,176,178,180,182,184,186,188,190,192,194,196,198,200,202,204,206,208,210,212,214,216,217) have mixed types. Specify dtype option on import or set low_memory=False.
  df_master = pd.read_csv("GSOY_Master_File.csv")


Unnamed: 0      STATION    DATE  LATITUDE  LONGITUDE  ELEVATION  \
0                 0  USC00401946  1949.0  35.51667  -85.28333      264.9   
1                 1  USC00401946  1950.0  35.51667  -85.28333      264.9   
2                 2  USC00401946  1951.0  35.51667  -85.28333      264.9   
3                 3  USC00401946  1952.0  35.51667  -85.28333      264.9   
4                 4  USC00401946  1953.0  35.51667  -85.28333      264.9   
...             ...          ...     ...       ...        ...        ...   
2525939     2525939  MXN00003093  2009.0  27.15000 -112.15000       19.8   
2525940     2525940  MXN00003093  2010.0  27.15000 -112.15000       19.8   
2525941     2525941  MXN00003093  2012.0  27.15000 -112.15000       19.8   
2525942     2525942  USC00055121  1949.0  40.24660 -105.14630     1569.7   
2525943     2525943  USC00055121  1950.0  40.24660 -105.14630     1569.7   

                            NAME   DP01 DP01_ATTRIBUTES   DP10  ...  \
0        COLLEGE JUNCTION, TN US  112.0               0   95.0  ...   
1        COLLEGE JUNCTION, TN US  127.0               0  105.0  ...   
2        COLLEGE JUNCTION, TN US  105.0               0   85.0  ...   
3        COLLEGE JUNCTION, TN US   91.0               0   75.0  ...   
4        COLLEGE JUNCTION, TN US  100.0               0   75.0  ...   
...                          ...    ...             ...    ...  ...   
2525939            SAN BRUNO, MX   13.0               m    7.0  ...   
2525940            SAN BRUNO, MX    6.0               m    4.0  ...   
2525941            SAN BRUNO, MX    NaN             NaN    NaN  ...   
2525942     LONGMONT 6 NW, CO US   75.0               0   32.0  ...   
2525943     LONGMONT 6 NW, CO US   65.0               0   31.0  ...   

        MN09_ATTRIBUTES  MX05 MX05_ATTRIBUTES  MX06  MX06_ATTRIBUTES  MX07  \
0                   NaN   NaN             NaN   NaN              NaN   NaN   
1                   NaN   NaN             NaN   NaN              NaN   NaN   
2                   NaN   NaN             NaN   NaN              NaN   NaN   
3                   NaN   NaN             NaN   NaN              NaN   NaN   
4                   NaN   NaN             NaN   NaN              NaN   NaN   
...                 ...   ...             ...   ...              ...   ...   
2525939             NaN   NaN             NaN   NaN              NaN   NaN   
2525940             NaN   NaN             NaN   NaN              NaN   NaN   
2525941             NaN   NaN             NaN   NaN              NaN   NaN   
2525942             NaN   NaN             NaN   NaN              NaN   NaN   
2525943             NaN   NaN             NaN   NaN              NaN   NaN   

        MX07_ATTRIBUTES  MX08 MX08_ATTRIBUTES  \
0                   NaN   NaN             NaN   
1                   NaN   NaN             NaN   
2                   NaN   NaN             NaN   
3                   NaN   NaN             NaN   
4                   NaN   NaN             NaN   
...                 ...   ...             ...   
2525939             NaN   NaN             NaN   
2525940             NaN   NaN             NaN   
2525941             NaN   NaN             NaN   
2525942             NaN   NaN             NaN   
2525943             NaN   NaN             NaN   

         The FTP site ftp://ftp.nodc.noaa.gov/ has been decommissioned  
0                                                      NaN              
1                                                      NaN              
2                                                      NaN              
3                                                      NaN              
4                                                      NaN              
...                                                    ...              
2525939                                                NaN              
2525940                                                NaN              
2525941                                                NaN  

### Explore the csv 

In [152]:
# List of column names
col_names  = [x for x in df_master.columns]
col_names

#Define metric of interest and metric of interest attributes
moi = 'TAVG'
moia = 'TAVG_ATTRIBUTES'

#Find proportion of stations that report the moi and the moia
moi_groupby = df_master.groupby(df_master.STATION)[moi].sum()
proportion_non_na_moi = round(len(moi_groupby[moi_groupby > 0])/len(moi_groupby)*100,2)
print(f"{proportion_non_na_moi}% of the stations have {moi} data")

moia_groupby = df_master.groupby(df_master.STATION)[moia].sum()
proportion_non_na_moia = round(len(moia_groupby[moia_groupby != 0])/len(moia_groupby)*100,2)
print(f"{proportion_non_na_moia}% of the stations have {moia} data")

37.3% of the stations have TAVG data
39.08% of the stations have TAVG_ATTRIBUTES data


#### Assume for now that TAVG_ATTRIBUTES cover all relevant TAVG data

## Subset the data to only data with the relevant MOI, then filter by date

In [153]:
#Make a list of stations
station_list_with_moi = moi_groupby[moi_groupby > 0].index.tolist()
#Subset the dataframe
df_moi = df_master[df_master['STATION'].isin(station_list_with_moi)]

In [154]:
#Convert data column to datetime dtype after converting to int
df_moi['DATE'] = df_moi['DATE'].astype(int)
df_moi['DATE'] = pd.to_datetime(df_moi['DATE'], format = '%Y')

C:\Users\dminkler002\AppData\Local\Temp\ipykernel_3936\4080105068.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_moi['DATE'] = df_moi['DATE'].astype(int)
C:\Users\dminkler002\AppData\Local\Temp\ipykernel_3936\4080105068.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_moi['DATE'] = pd.to_datetime(df_moi['DATE'], format = '%Y')


In [155]:
#What is the most recent date?
max(df_moi['DATE']) #2022

#Subset DF and Check to see how many stations contain a data range of 20 years 2002 to 2022
df_moi_date = df_moi[(df_moi['DATE'] >= datetime(2002,1,1)) & (df_moi['DATE'] <= datetime(2022,1,1))]
print(round(len(df_moi_date['STATION'].unique())/len(df_master['STATION'].unique())*100),2)

#Check for missing dates b/t 2002 and 2022 


23 2


In [156]:
full_date_range =  [x.year for x in pd.date_range(start = '1/1/2002', freq = 'YS', periods = 20)] # YS = year start 

# Pick a random station to test the date range approach to check for missing values
random.seed(20)
station_list = df_moi_date['STATION'].unique()
station_dates = [x.year for x in df_moi_date.loc[df_moi_date['STATION'] == random.choice(station_list),]['DATE']]

# Identify all station names which satisfy this date range

station_list_daterange = []
for i in station_list:
    station_dates = [x.year for x in df_moi_date.loc[df_moi_date['STATION'] == i,]['DATE']]
    date_check = [y for y in full_date_range if y not in station_dates]
    if len(date_check) == 0:
        station_list_daterange.append(i)

In [157]:
len(station_list_daterange)

4225

#### If we choose a date range of 2002 to 2022, we get 1441 stations which meet and have the required dates, If we choose a date range of 2002 to 2021, we get 4225 stations which meet and have the required dates
### We will go with a date range of 2002 to 2021. Subset the df by the station list where criteria has been met

In [158]:
# Code to subset by station list 
df_moi_date_range_20y = df_moi_date.loc[df_moi_date['STATION'].isin(station_list_daterange), ]

#Filter out any date values greater than 2021
df_moi_date_range_20y = df_moi_date_range_20y[df_moi_date_range_20y['DATE'] < datetime(2022,1,1)]

#Sanity check by checking group size of all stations (should be zero)
len([x for x in df_moi_date_range_20y.groupby('STATION').size() if x != 20])


0

In [159]:
df_moi_date_range_20y

Unnamed: 0      STATION       DATE  LATITUDE  LONGITUDE  ELEVATION  \
170             170  USW00003822 2002-01-01  32.13133   -81.2023        8.7   
171             171  USW00003822 2003-01-01  32.13133   -81.2023        8.7   
172             172  USW00003822 2004-01-01  32.13133   -81.2023        8.7   
173             173  USW00003822 2005-01-01  32.13133   -81.2023        8.7   
174             174  USW00003822 2006-01-01  32.13133   -81.2023        8.7   
...             ...          ...        ...       ...        ...        ...   
2524916     2524916  GME00121978 2017-01-01  51.40560     6.9683      150.0   
2524917     2524917  GME00121978 2018-01-01  51.40560     6.9683      150.0   
2524918     2524918  GME00121978 2019-01-01  51.40560     6.9683      150.0   
2524919     2524919  GME00121978 2020-01-01  51.40560     6.9683      150.0   
2524920     2524920  GME00121978 2021-01-01  51.40560     6.9683      150.0   

                                          NAME   DP01 DP01_ATTRIBUTES   DP10  \
170      SAVANNAH INTERNATIONAL AIRPORT, GA US  117.0               0   75.0   
171      SAVANNAH INTERNATIONAL AIRPORT, GA US  134.0               0   76.0   
172      SAVANNAH INTERNATIONAL AIRPORT, GA US  112.0               0   63.0   
173      SAVANNAH INTERNATIONAL AIRPORT, GA US  117.0               0   70.0   
174      SAVANNAH INTERNATIONAL AIRPORT, GA US   97.0               0   59.0   
...                                        ...    ...             ...    ...   
2524916                     ESSEN BREDENEY, GM  188.0               E  105.0   
2524917                     ESSEN BREDENEY, GM  144.0               E   81.0   
2524918                     ESSEN BREDENEY, GM  172.0               E  110.0   
2524919                     ESSEN BREDENEY, GM  166.0               E   92.0   
2524920                     ESSEN BREDENEY, GM  187.0               E  104.0   

         ... MN09_ATTRIBUTES  MX05 MX05_ATTRIBUTES  MX06  MX06_ATTRIBUTES  \
170      ...             NaN   NaN             NaN   NaN              NaN   
171      ...             NaN   NaN             NaN   NaN              NaN   
172      ...             NaN   NaN             NaN   NaN              NaN   
173      ...             NaN   NaN             NaN   NaN              NaN   
174      ...             NaN   NaN             NaN   NaN              NaN   
...      ...             ...   ...             ...   ...              ...   
2524916  ...             NaN   NaN             NaN   NaN              NaN   
2524917  ...             NaN   NaN             NaN   NaN              NaN   
2524918  ...             NaN   NaN             NaN   NaN              NaN   
2524919  ...             NaN   NaN             NaN   NaN              NaN   
2524920  ...             NaN   NaN             NaN   NaN              NaN   

         MX07 MX07_ATTRIBUTES  MX08 MX08_ATTRIBUTES  \
170       NaN             NaN   NaN             NaN   
171       NaN             NaN   NaN             NaN   
172       NaN             NaN   NaN             NaN   
173       NaN             NaN   NaN             NaN   
174       NaN             NaN   NaN             NaN   
...       ...             ...   ...             ...   
2524916   NaN             NaN   NaN             NaN   
2524917   NaN             NaN   NaN             NaN   
2524918   NaN             NaN   NaN             NaN   
2524919   NaN             NaN   NaN             NaN   
2524920   NaN             NaN   NaN             NaN   

         The FTP site ftp://ftp.nodc.noaa.gov/ has been decommissioned  
170                                                    NaN              
171                                                    NaN              
172                                                    NaN              
173                                                    NaN              
174                                                    NaN              
...                                                    ...              
25

## Extract Country Codes and Match with Country Names

In [160]:
# Extract the country code from the GHCN station ID using REGEX, create new columns. Extract 2 and 3 character codes 
df_moi_date_range_20y['COUNTRY_CODE_2_CHAR'] = df_moi_date_range_20y['STATION'].str.extract(r'(\D{2})')
df_moi_date_range_20y['COUNTRY_CODE_3_CHAR'] = df_moi_date_range_20y['STATION'].str.extract(r'(\D+)')

In [161]:
# Explore the country codes found (extracted from GHCN Station ID)
df_moi_date_range_20y['COUNTRY_CODE_2_CHAR'].unique().tolist()

print('There are {} country codes in our data set. These include {}'.format(
    len(df_moi_date_range_20y['COUNTRY_CODE_2_CHAR'].unique()),
    df_moi_date_range_20y['COUNTRY_CODE_2_CHAR'].unique().tolist()
        )
     )

There are 32 country codes in our data set. These include ['US', 'GM', 'SW', 'NO', 'FI', 'AS', 'SP', 'AU', 'UK', 'FR', 'FM', 'NL', 'SZ', 'CH', 'VQ', 'EN', 'GQ', 'SI', 'RI', 'IS', 'CA', 'BU', 'RO', 'HR', 'AG', 'DA', 'RM', 'RQ', 'LO', 'SF', 'BK', 'JA']


In [162]:
# Read text file with country code legend from https://www.ncei.noaa.gov/products/land-based-station/global-historical-climatology-network-daily
with open("ghcn_country_codes.txt") as f:
    contents = f.readlines()

# Remove \n in all of the strings 
contents = [x.replace(' \n', '').replace('\n', '') for x in contents]

# Convert list to a dictionary 
country_names = {x[0:2]:x[3:] for x in contents}

# Convert Dictionary to pandas dataframe to make joining easier
country_names = pd.DataFrame.from_dict(country_names, orient = 'index', columns = ['country_name'])
country_names.head()




country_name
AC   Antigua and Barbuda
AE  United Arab Emirates
AF           Afghanistan
AG               Algeria
AJ            Azerbaijan

In [164]:
set(country_names.columns) - set(df_moi_date_range_20y.columns)

{'country_name'}

In [165]:
# Join Country Names to DataFrame
df_moi_date_range_20y = df_moi_date_range_20y.merge(country_names, how='left', left_on = 'COUNTRY_CODE_2_CHAR', right_on = country_names.index)

df_moi_date_range_20y.head(5)

Unnamed: 0      STATION       DATE  LATITUDE  LONGITUDE  ELEVATION  \
0         170  USW00003822 2002-01-01  32.13133   -81.2023        8.7   
1         171  USW00003822 2003-01-01  32.13133   -81.2023        8.7   
2         172  USW00003822 2004-01-01  32.13133   -81.2023        8.7   
3         173  USW00003822 2005-01-01  32.13133   -81.2023        8.7   
4         174  USW00003822 2006-01-01  32.13133   -81.2023        8.7   

                                    NAME   DP01 DP01_ATTRIBUTES  DP10  ...  \
0  SAVANNAH INTERNATIONAL AIRPORT, GA US  117.0               0  75.0  ...   
1  SAVANNAH INTERNATIONAL AIRPORT, GA US  134.0               0  76.0  ...   
2  SAVANNAH INTERNATIONAL AIRPORT, GA US  112.0               0  63.0  ...   
3  SAVANNAH INTERNATIONAL AIRPORT, GA US  117.0               0  70.0  ...   
4  SAVANNAH INTERNATIONAL AIRPORT, GA US   97.0               0  59.0  ...   

  MX06  MX06_ATTRIBUTES MX07  MX07_ATTRIBUTES  MX08  MX08_ATTRIBUTES  \
0  NaN              NaN  NaN              NaN   NaN              NaN   
1  NaN              NaN  NaN              NaN   NaN              NaN   
2  NaN              NaN  NaN              NaN   NaN              NaN   
3  NaN              NaN  NaN              NaN   NaN              NaN   
4  NaN              NaN  NaN              NaN   NaN              NaN   

  The FTP site ftp://ftp.nodc.noaa.gov/ has been decommissioned  \
0                                                NaN              
1                                                NaN              
2                                                NaN              
3                                                NaN              
4                                                NaN              

   COUNTRY_CODE_2_CHAR COUNTRY_CODE_3_CHAR   country_name  
0                   US                 USW  United States  
1                   US                 USW  United States  
2                   US                 USW  United States  
3                   US                 USW  United States  
4                   US                 USW  United States  

[5 rows x 221 columns]

## Join Reverse Geocoding data to DF

In [167]:
#Load in lat-lon df from geocoding, drop unnamed column
lat_lon_df = pd.read_csv('lat_lon_df.csv').drop('Unnamed: 0', axis = 1)

## Need to convert a dictionary which is saved as a string to an actual dictionary
## First, convert nan to 'nan' so that eval func. can be used

#Test case

#regex = re.compile('\s(nan)(?=\W)')
#print(regex.search(lat_lon_df['REVERSE_GEOCODE'][1]))

# Final solution 
lat_lon_df['REVERSE_GEOCODE'] = lat_lon_df['REVERSE_GEOCODE'].str.replace(r'\s(nan)(?=\W)', "'nan'", regex = True)

## Then convert strings to dicts
lat_lon_df['REVERSE_GEOCODE'] = lat_lon_df['REVERSE_GEOCODE'].apply(eval)

# Unpack the REVERSE_GEOCODE column to multiple columns, then join to main 20y_sub df
lat_lon_df = pd.concat([lat_lon_df.loc[:,'lat_lon'], lat_lon_df['REVERSE_GEOCODE'].apply(pd.Series)], axis=1)

lat_lon_df

lat_lon           city                state postcode  \
0       32.13133,-81.2023       Savannah              Georgia    31418   
1           39.68,-111.32            nan                 Utah    84629   
2       39.0055,-114.2205            nan               Nevada      nan   
3     37.62544,-120.95492            nan           California    95319   
4         39.098,-79.4322            nan        West Virginia    26260   
...                   ...            ...                  ...      ...   
4219   44.35904,-89.83695            nan            Wisconsin    54494   
4220    -37.8333,140.7833  Mount Gambier      South Australia     5290   
4221        48.13,-115.62            nan              Montana      nan   
4222   55.22068,-162.7323       Cold Bay               Alaska    99571   
4223       51.4056,6.9683          Essen  Nordrhein-Westfalen    45133   

                 county region  
0        Chatham County    nan  
1        Sanpete County    nan  
2     White Pine County    nan  
3     Stanislaus County    nan  
4         Tucker County    nan  
...                 ...    ...  
4219        Wood County    nan  
4220                nan    nan  
4221     Lincoln County    nan  
4222     Aleutians East    nan  
4223                nan    nan  

[4224 rows x 6 columns]

In [168]:
# Create a lat_lon column which is a concatenated latitude and longitude 
df_moi_date_range_20y = df_moi_date_range_20y\
.assign(lat_lon = lambda df: (df.LATITUDE.astype(str) + "," + df.LONGITUDE.astype(str)))

# Join df_moi_date_range_20y with the lat_lon_df from the reverse geocoding 
df_moi_date_range_20y = df_moi_date_range_20y.merge(lat_lon_df\
                                                    .loc[:, ['lat_lon','city','state', 'postcode', 'county', 'region']], 
                                                    how='left', on='lat_lon')

# Check out the df
df_moi_date_range_20y
                                                    

Unnamed: 0      STATION       DATE  LATITUDE  LONGITUDE  ELEVATION  \
0             170  USW00003822 2002-01-01  32.13133   -81.2023        8.7   
1             171  USW00003822 2003-01-01  32.13133   -81.2023        8.7   
2             172  USW00003822 2004-01-01  32.13133   -81.2023        8.7   
3             173  USW00003822 2005-01-01  32.13133   -81.2023        8.7   
4             174  USW00003822 2006-01-01  32.13133   -81.2023        8.7   
...           ...          ...        ...       ...        ...        ...   
84495     2524916  GME00121978 2017-01-01  51.40560     6.9683      150.0   
84496     2524917  GME00121978 2018-01-01  51.40560     6.9683      150.0   
84497     2524918  GME00121978 2019-01-01  51.40560     6.9683      150.0   
84498     2524919  GME00121978 2020-01-01  51.40560     6.9683      150.0   
84499     2524920  GME00121978 2021-01-01  51.40560     6.9683      150.0   

                                        NAME   DP01 DP01_ATTRIBUTES   DP10  \
0      SAVANNAH INTERNATIONAL AIRPORT, GA US  117.0               0   75.0   
1      SAVANNAH INTERNATIONAL AIRPORT, GA US  134.0               0   76.0   
2      SAVANNAH INTERNATIONAL AIRPORT, GA US  112.0               0   63.0   
3      SAVANNAH INTERNATIONAL AIRPORT, GA US  117.0               0   70.0   
4      SAVANNAH INTERNATIONAL AIRPORT, GA US   97.0               0   59.0   
...                                      ...    ...             ...    ...   
84495                     ESSEN BREDENEY, GM  188.0               E  105.0   
84496                     ESSEN BREDENEY, GM  144.0               E   81.0   
84497                     ESSEN BREDENEY, GM  172.0               E  110.0   
84498                     ESSEN BREDENEY, GM  166.0               E   92.0   
84499                     ESSEN BREDENEY, GM  187.0               E  104.0   

       ... The FTP site ftp://ftp.nodc.noaa.gov/ has been decommissioned  \
0      ...                                                NaN              
1      ...                                                NaN              
2      ...                                                NaN              
3      ...                                                NaN              
4      ...                                                NaN              
...    ...                                                ...              
84495  ...                                                NaN              
84496  ...                                                NaN              
84497  ...                                                NaN              
84498  ...                                                NaN              
84499  ...                                                NaN              

       COUNTRY_CODE_2_CHAR COUNTRY_CODE_3_CHAR   country_name  \
0                       US                 USW  United States   
1                       US                 USW  United States   
2                       US                 USW  United States   
3                       US                 USW  United States   
4                       US                 USW  United States   
...                    ...                 ...            ...   
84495                   GM                 GME        Germany   
84496                   GM                 GME        Germany   
84497                   GM                 GME        Germany   
84498                   GM                 GME        Germany   
84499                   GM                 GME        Germany   

                 lat_lon      city                state  postcode  \
0      32.13133,-81.2023  Savannah              Georgia     31418   
1      32.13133,-81.2023  Savannah              Georgia     31418   
2      32.13133,-81.2023  Savannah              Georgia     31418   
3      32.13133,-81.2023  Savannah              Georgia     31418   
4      32.13133,-81.2023  Savannah              Georgia     31418   
...                  ...       ...   

## Define US Regions: Northeast, Southeast, Midwest, Southwest, Pacific Northwest

In [169]:
# Create dictionary 
us_regions = {'Northeast': ['Maine', 'New Hampshire', 'Vermont', 'Massachusetts', 'Connecticut', 'New York', 
                            'New Jersey', 'Pennsylvania', 'Maryland','Rhode Island', 'Delaware'],
             'Southeast': ['Virginia', 'West Virginia', 'Tennessee', 'North Carolina', 'South Carolina',
                           'Georgia', 'Florida', 'Alabama', 'Mississippi', 'Louisiana', 'Kentucky', 'Arkansas'],
             'Southwest': ['Arizona', 'New Mexico', 'Texas', 'Oklahoma', 'Nevada', 'Colorado', 'Utah', 'California'],
             'Midwest': ['North Dakota', 'South Dakota', 'Nebraska', 'Kansas', 'Minnesota', 'Iowa', 
                        'Missouri', 'Wisconsin', 'Illinois', 'Michigan', 'Indiana', 'Ohio'],
             'Northwest': ['Washington', 'Oregon', 'Idaho','Montana', 'Wyoming']}

#Convert dictionary to a usable dataframe. 
temp_df = pd.DataFrame(us_regions.values(), index=us_regions.keys()).reset_index()
us_regions_df = temp_df.melt(id_vars=['index'])
us_regions_df = us_regions_df.loc[~us_regions_df['value'].isna() , ['index', 'value']].reset_index(drop = True)\
                .sort_values('index')

us_regions_df.columns = ['cont_us_region', 'state']

# Test to see if any states were missed
set(df_moi_date_range_20y.loc[df_moi_date_range_20y['COUNTRY_CODE_2_CHAR'] == 'US', 'state'].unique()).difference(set(us_regions_df['state']))
    

{'Alaska', 'Hawaii'}

In [171]:
# Join to main df
df_moi_date_range_20y = \
df_moi_date_range_20y\
.merge(us_regions_df, how='left', on = 'state')

df_moi_date_range_20y.head(5)

Unnamed: 0      STATION       DATE  LATITUDE  LONGITUDE  ELEVATION  \
0         170  USW00003822 2002-01-01  32.13133   -81.2023        8.7   
1         171  USW00003822 2003-01-01  32.13133   -81.2023        8.7   
2         172  USW00003822 2004-01-01  32.13133   -81.2023        8.7   
3         173  USW00003822 2005-01-01  32.13133   -81.2023        8.7   
4         174  USW00003822 2006-01-01  32.13133   -81.2023        8.7   

                                    NAME   DP01 DP01_ATTRIBUTES  DP10  ...  \
0  SAVANNAH INTERNATIONAL AIRPORT, GA US  117.0               0  75.0  ...   
1  SAVANNAH INTERNATIONAL AIRPORT, GA US  134.0               0  76.0  ...   
2  SAVANNAH INTERNATIONAL AIRPORT, GA US  112.0               0  63.0  ...   
3  SAVANNAH INTERNATIONAL AIRPORT, GA US  117.0               0  70.0  ...   
4  SAVANNAH INTERNATIONAL AIRPORT, GA US   97.0               0  59.0  ...   

  COUNTRY_CODE_2_CHAR  COUNTRY_CODE_3_CHAR   country_name            lat_lon  \
0                  US                  USW  United States  32.13133,-81.2023   
1                  US                  USW  United States  32.13133,-81.2023   
2                  US                  USW  United States  32.13133,-81.2023   
3                  US                  USW  United States  32.13133,-81.2023   
4                  US                  USW  United States  32.13133,-81.2023   

       city    state postcode          county region  cont_us_region  
0  Savannah  Georgia    31418  Chatham County    nan       Southeast  
1  Savannah  Georgia    31418  Chatham County    nan       Southeast  
2  Savannah  Georgia    31418  Chatham County    nan       Southeast  
3  Savannah  Georgia    31418  Chatham County    nan       Southeast  
4  Savannah  Georgia    31418  Chatham County    nan       Southeast  

[5 rows x 228 columns]

## Export to CSV for Tableau

In [173]:
# Subset Columns to only relevant columns 

column_subset = ['STATION', 'DATE', 'LATITUDE', 'LONGITUDE', 'ELEVATION', 'NAME', 
                 moi, moia, 'COUNTRY_CODE_2_CHAR', 'country_name',
                'city', 'state', 'postcode', 'county', 'cont_us_region']
df_moi_date_range_20y_sub = df_moi_date_range_20y.loc[:, column_subset]
df_moi_date_range_20y_sub.head()

STATION       DATE  LATITUDE  LONGITUDE  ELEVATION  \
0  USW00003822 2002-01-01  32.13133   -81.2023        8.7   
1  USW00003822 2003-01-01  32.13133   -81.2023        8.7   
2  USW00003822 2004-01-01  32.13133   -81.2023        8.7   
3  USW00003822 2005-01-01  32.13133   -81.2023        8.7   
4  USW00003822 2006-01-01  32.13133   -81.2023        8.7   

                                    NAME   TAVG TAVG_ATTRIBUTES  \
0  SAVANNAH INTERNATIONAL AIRPORT, GA US  19.53               0   
1  SAVANNAH INTERNATIONAL AIRPORT, GA US  19.10               0   
2  SAVANNAH INTERNATIONAL AIRPORT, GA US  19.02               0   
3  SAVANNAH INTERNATIONAL AIRPORT, GA US  18.84               0   
4  SAVANNAH INTERNATIONAL AIRPORT, GA US  19.18               0   

  COUNTRY_CODE_2_CHAR   country_name      city    state postcode  \
0                  US  United States  Savannah  Georgia    31418   
1                  US  United States  Savannah  Georgia    31418   
2                  US  United States  Savannah  Georgia    31418   
3                  US  United States  Savannah  Georgia    31418   
4                  US  United States  Savannah  Georgia    31418   

           county cont_us_region  
0  Chatham County      Southeast  
1  Chatham County      Southeast  
2  Chatham County      Southeast  
3  Chatham County      Southeast  
4  Chatham County      Southeast

In [175]:
# Export Subset to CSV to explore in Tableau 
df_moi_date_range_20y_sub.to_csv('df_moi_date_range_20y_sub.csv')

# Linear Regression for US States. 

## Preprocessing (convert years, subset DF)

In [10]:
# Read in CSV
df_moi_date_range_20y_sub = pd.read_csv('../df_moi_date_range_20y_sub.csv')

#See CSV and Columns
df_moi_date_range_20y_sub.head(5)

Unnamed: 0      STATION        DATE  LATITUDE  LONGITUDE  ELEVATION  \
0           0  USW00003822  2002-01-01  32.13133   -81.2023        8.7   
1           1  USW00003822  2003-01-01  32.13133   -81.2023        8.7   
2           2  USW00003822  2004-01-01  32.13133   -81.2023        8.7   
3           3  USW00003822  2005-01-01  32.13133   -81.2023        8.7   
4           4  USW00003822  2006-01-01  32.13133   -81.2023        8.7   

                                    NAME   TAVG TAVG_ATTRIBUTES  \
0  SAVANNAH INTERNATIONAL AIRPORT, GA US  19.53               0   
1  SAVANNAH INTERNATIONAL AIRPORT, GA US  19.10               0   
2  SAVANNAH INTERNATIONAL AIRPORT, GA US  19.02               0   
3  SAVANNAH INTERNATIONAL AIRPORT, GA US  18.84               0   
4  SAVANNAH INTERNATIONAL AIRPORT, GA US  19.18               0   

  COUNTRY_CODE_2_CHAR   country_name region_x       subregion      city  \
0                  US  United States       GA  Chatham County  Savannah   
1                  US  United States       GA  Chatham County  Savannah   
2                  US  United States       GA  Chatham County  Savannah   
3                  US  United States       GA  Chatham County  Savannah   
4                  US  United States       GA  Chatham County  Savannah   

     state postcode          county region_y cont_us_region  
0  Georgia    31418  Chatham County      NaN      Southeast  
1  Georgia    31418  Chatham County      NaN      Southeast  
2  Georgia    31418  Chatham County      NaN      Southeast  
3  Georgia    31418  Chatham County      NaN      Southeast  
4  Georgia    31418  Chatham County      NaN      Southeast

In [44]:
# Define start year and create new column for years since 2002
years_since = 2002

# Create a dynamic time difference column
df_moi_date_range_20y_sub['Years since ' + str(years_since)] = pd.to_datetime(df_moi_date_range_20y_sub['DATE']) - datetime(2002,1,1)

# Convert time difference to years (from days)
df_moi_date_range_20y_sub['Years since ' + str(years_since)] = df_moi_date_range_20y_sub['Years since ' + str(years_since)].astype("timedelta64[Y]")


0         0.0
1         0.0
2         1.0
3         3.0
4         4.0
         ... 
84495    15.0
84496    16.0
84497    16.0
84498    17.0
84499    19.0
Name: Years since 2002, Length: 84500, dtype: float64

# (Long run time) Add state, and sub-region information to dataframe using Nominatim (geocoder)

In [ ]:
# Test Case Using a US Lat/Lon

from geopy.geocoders import Nominatim
geolocator = Nominatim(user_agent = "Global Summary of the Year Analysis")
location = geolocator.reverse("51.40, 6.9683")
location.raw


# Test Case Using a German Lat/Lon

from geopy.geocoders import Nominatim
geolocator = Nominatim(user_agent = "Global Summary of the Year Analysis")
location = geolocator.reverse("32.13, -81.20").raw['address']['city']
location # Test case works!

In [ ]:
# Define a function that returns city, state, postcode, county, and region, if avaiable,
# based on the lat_lon of a station

def reverse_geocode(lat_lon):
    location_details = {'city': np.nan, 'state': np.nan, 'postcode': np.nan, 'county': np.nan, 'region': np.nan}
    geolocator = Nominatim(user_agent = "Global Summary of the Year Analysis")
    location = geolocator.reverse(lat_lon, timeout = 600) #Allows for 10 min, or 600 seconds before timing out
    
    if 'raw' in dir(location):
        location = location.raw['address']
        
        if 'city' in location.keys():
            location_details['city'] = location['city']

        if 'state' in location.keys():
            location_details['state'] = location['state']

        if 'postcode' in location.keys():
            location_details['postcode'] = location['postcode']
        
        if 'county' in location.keys():
            location_details['county'] = location['county']           
          
        if 'region' in location.keys():
            location_details['region'] = location['region']
        
    return location_details

In [ ]:
# Sample subset of rows and test reverse_geocode function

from tqdm._tqdm_notebook import tqdm_notebook
tqdm_notebook.pandas()
import swifter

df_moi_date_range_20y_sample = df_moi_date_range_20y.sample(n=250)


df_moi_date_range_20y_sample['CITY'] = df_moi_date_range_20y_sample['lat_lon'].swifter.apply(lambda x: reverse_geocode(x)['city'])
# df_moi_date_range_20y_sample['STATE'] = df_moi_date_range_20y_sample['lat_lon'].apply(lambda x: reverse_geocode(x)['state'])
# df_moi_date_range_20y_sample['POSTCODE'] = df_moi_date_range_20y_sample['lat_lon'].apply(lambda x: reverse_geocode(x)['postcode'])
# df_moi_date_range_20y_sample['COUNTY'] = df_moi_date_range_20y_sample['lat_lon'].apply(lambda x: reverse_geocode(x)['county'])
# df_moi_date_range_20y_sample['REGION'] = df_moi_date_range_20y_sample['lat_lon'].apply(lambda x: reverse_geocode(x)['region'])

#Vectorize geo_location: (this failed)
#df_moi_date_range_20y_sample['CITY'] = reverse_geocode(df_moi_date_range_20y_sample['lat_lon'])['city']




In [ ]:
# Create funtion that adds city, state, postcode, county, region to a dataframe
# Function requires a column 'lat_lon'

def cspcr(dataframe):
    dataframe['CITY'] = dataframe['lat_lon'].progress_apply(lambda x: reverse_geocode(x)['city'])
    dataframe['STATE'] = dataframe['lat_lon'].progress_apply(lambda x: reverse_geocode(x)['state'])
    dataframe['POSTCODE'] = dataframe['lat_lon'].progress_apply(lambda x: reverse_geocode(x)['postcode'])
    dataframe['COUNTY'] = dataframe['lat_lon'].progress_apply(lambda x: reverse_geocode(x)['county'])
    dataframe['REGION'] = dataframe['lat_lon'].progress_apply(lambda x: reverse_geocode(x)['region'])
    
    return dataframe

#Alternative approach, append entire dictionary to dataframe, then unpack later to save number of apply calls 
def efficient_cspcr(dataframe):
    dataframe['REVERSE_GEOCODE'] = dataframe['lat_lon'].progress_apply(lambda x: reverse_geocode(x))
    
    return dataframe

In [ ]:
# Use function to get city, state, postcode, county, and region added to temporary lat_lon df

lat_lon_df = pd.DataFrame({'lat_lon': df_moi_date_range_20y['lat_lon'].unique()})

lat_lon_df = efficient_cspcr(lat_lon_df)

lat_lon_df.head(5)

### Write Lat_Lon Geocode Data to CSV

In [ ]:
# Write lat_lon_df to CSV to avoid having to do that again! (took 38 min)

lat_lon_df.to_csv("lat_lon_df.csv")


# Appendix

In [182]:
#### Appendix: Troubleshooting NAs for Postcode/State/City

# Check how many rows have an NA for Postcode + State + City and see if there are any commonalities 

nada = df_moi_date_range_20y_sample\
    .loc[df_moi_date_range_20y_sample\
    .loc[:,['CITY', 'STATE','POSTCODE', 'COUNTY', 'REGION']]\
    .isnull().all(axis = 1) , ['STATION', 'lat_lon']]

#for i in nada.lat_lon:
    #print(geolocator.reverse(i).raw['address'])
    
# Based on this, add county and region to the function 

len(nada)

0

In [22]:
#### Appendix: This approach had too much missing data. Used reverse geocoding instead.


# Read in master location idenfitier database (mlid)
mlid = pd.read_excel('master-location-identifier-database.xlsx', header = 4)


# Join on station code (ghcn) and bring in redion and subregion
df_moi_date_range_20y = df_moi_date_range_20y.merge(mlid.loc[:, ['ghcn','region', 'subregion']] , 
                                                    how='left', left_on='STATION', right_on='ghcn')
                                                    

# What percent of data has region and subregion information?

print('For all countries')
regions = ['region', 'subregion']
for i in regions:
    percent_missing(df_moi_date_range_20y, 'STATION', i)
    
# What percent of US data has region and subregion information?
print('For just US')
regions = ['region', 'subregion']
for i in regions:
    temp_df = df_moi_date_range_20y.loc[(df_moi_date_range_20y['COUNTRY_CODE_2_CHAR'] == 'US'),:]
    percent_missing(temp_df, 'STATION', i)
    
# What percent of NON-US data has region and subregion information?
print('For everywhere but US')
regions = ['region', 'subregion']
for i in regions:
    temp_df = df_moi_date_range_20y.loc[(df_moi_date_range_20y['COUNTRY_CODE_2_CHAR'] != 'US'),:]
    percent_missing(temp_df, 'STATION', i)



country3 country2      country region subregion  \
0       ABW       AW        Aruba      -         -   
1       ABW       AW        Aruba     SA         -   
2       AFG       AF  Afghanistan    BAL         -   
3       AFG       AF  Afghanistan    BAL         -   
4       AFG       AF  Afghanistan    BAL         -   
5       AFG       AF  Afghanistan    BAL         -   
6       AFG       AF  Afghanistan    BAM         -   
7       AFG       AF  Afghanistan    BAM         -   
8       AFG       AF  Afghanistan    BAM         -   
9       AFG       AF  Afghanistan    BAM         -   
10      AFG       AF  Afghanistan    BDG         -   
11      AFG       AF  Afghanistan    BDG         -   
12      AFG       AF  Afghanistan    BDG         -   
13      AFG       AF  Afghanistan    BDG         -   
14      AFG       AF  Afghanistan    BDG         -   
15      AFG       AF  Afghanistan    BDG         -   
16      AFG       AF  Afghanistan    BDS         -   
17      AFG       AF  Afghanistan    BDS         -   
18      AFG       AF  Afghanistan    BDS         -   
19      AFG       AF  Afghanistan    BDS         -   
20      AFG       AF  Afghanistan    BDS         -   
21      AFG       AF  Afghanistan    BDS         -   
22      AFG       AF  Afghanistan    BDS         -   
23      AFG       AF  Afghanistan    BDS         -   
24      AFG       AF  Afghanistan    BDS         -   
25      AFG       AF  Afghanistan    BDS         -   
26      AFG       AF  Afghanistan    BDS         -   
27      AFG       AF  Afghanistan    BDS         -   
28      AFG       AF  Afghanistan    BGL         -   
29      AFG       AF  Afghanistan    BGL         -   
30      AFG       AF  Afghanistan    BGL         -   
31      AFG       AF  Afghanistan    BGL         -   
32      AFG       AF  Afghanistan    FRA         -   
33      AFG       AF  Afghanistan    FRA         -   
34      AFG       AF  Afghanistan    FRA         -   
35      AFG       AF  Afghanistan    FRA         -   
36      AFG       AF  Afghanistan    FRA         -   
37      AFG       AF  Afghanistan    FRA         -   
38      AFG       AF  Afghanistan    FRA         -   
39      AFG       AF  Afghanistan    FYB         -   
40      AFG       AF  Afghanistan    FYB         -   
41      AFG       AF  Afghanistan    FYB         -   
42      AFG       AF  Afghanistan    FYB         -   
43      AFG       AF  Afghanistan    FYB         -   
44      AFG       AF  Afghanistan    GHA         -   
45      AFG       AF  Afghanistan    GHA         -   
46      AFG       AF  Afghanistan    GHA         -   
47      AFG       AF  Afghanistan    GHA         -   
48      AFG       AF  Afghanistan    GHA         -   
49      AFG       AF  Afghanistan    GHA         -   

                                           place_name  \
0                                         Old Norwood   
1                                   Upper Hell's Gate   
2                  Kyzylabad|Mazar-I-Sharif|Qizil?b?d   
3                    Pushti-Bag|Mazari Sharif|Dehd?d?   
4             Qala-i-Gul Mohd|Mazari Sharif|Qizil?b?d   
5                                 Shor-Tipa|Shor Tepe   
6                 Dekhi-Khazara|Bamiyan|Deh-e Haz?rah   
7                                   Gorvana|Sar ?sy?b   
8                        Kala-Nau|Pan Jao|Dahan-e M?r   
9                           Tadzhik|Bamiyan PRT|T?jik   
10                         Bagak|Qal-E-Naw|Qala i Naw   
11                     Ch?kar?n|Qala-I-Naw|Qala i Naw   
12                                      K?bul? Far?r?   
13                      Kh??ad?r|Qurmach|Bala Murghab   
14                 Towraghondt|Towraghondi|T?ragh?n??   
15                              Zaysalekh|Qades|Q?dis   
16                     B?y M?las?|Feyzabad|B?? M?las?   
17                                      Feyzabad|K?r?   
18                   G?win Darah|Iskachem|G?wen Darah   
19                                       Keshm|Bal?ch   
20                                Kevron|Darwaz|N?s